### PR TITLE
Update travis CI to add gcc6 and gcc4.9

### PR DESCRIPTION
**Pull request by jbcoe**
Changes to travis adopted from @philipcraig's PR

### CODE REVIEW

**Comment by jbcoe** [comment]
Failing tests for gcc4.9.

Course of action:

Add `ifdef` checks.
Add CMake definitions to enable modern C++ for gcc 4.9.

**Comment by philipcraig** [comment]
I think @philcraig doesn't care about this PR. I do though!

File: .travis.yml [hunk 1]
```diff
@@ -23,7 +23,23 @@ matrix:
...
+    # 3/ Linux GCC 4.9 Builds
+    - os: linux
+      compiler: gcc
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-4.9
+            - g++-6
```
Comment (viboes) [inline .travis.yml]: Why version 6 also?

File: .travis.yml [hunk 2]
```diff
@@ -23,7 +23,23 @@ matrix:
           packages: ['g++-6']
       env: COMPILER='g++-6'
 
-#    # 3/ OSX Clang Builds - disabled for now
+    # 3/ Linux GCC 4.9 Builds
```
Comment (viboes) [inline .travis.yml]: version 4.9 is yet a Linux GCC Build, isn't it?

Comment (jbcoe) [reply .travis.yml]: I'm not sure I understand.

File: .travis.yml (hunk 1 above)
Comment (jbcoe) [reply .travis.yml]: Update alternatives is called for all Linux builds to select gcc-6.0. It needs to exist.

It is not necessarily selected as the C++ compiler.

There may be a better way to do this.

Comment (philipcraig) [reply .travis.yml]: It's what I had to do in my own travis to get 4.9 to install correctly

**Pull request closed without merging**